tests/build/test_generate_config.py

```diff
@@ -2,31 +2,294 @@
+
+    config_text = program.files.cmake_config_file.read_text()
+
+    assert target in config_text
+    assert toolchain in config_text
+
+
+def test_config_param_from_mbed_lib_processed_with_default_name_mangling(program):
+    create_mbed_lib_json(
+        program.mbed_os.root / "platform" / "mbed_lib.json",
+        "platform",
+        config={
+            "stdio-convert-newlines": {
+                "help": "Enable conversion to standard newlines on stdin/stdout/stderr",
+                "value": True,
+            }
+        },
+    )
+
+    generate_config("K64F", "GCC_ARM", program)
+
+    config_text = program.files.cmake_config_file.read_text()
+
+    assert "MBED_CONF_PLATFORM_STDIO_CONVERT_NEWLINES" in config_text
+
+
+def test_config_param_from_mbed_lib_processed_with_user_set_name(program):
+    create_mbed_lib_json(
+        program.mbed_os.root / "platform" / "mbed_lib.json",
+        "platform",
+        config={
+            "stdio-convert-newlines": {
+                "help": "Enable conversion to standard newlines on stdin/stdout/stderr",
+                "value": True,
+                "macro_name": "ENABLE_NEWLINES",
+            }
+        },
+    )
+
+    generate_config("K64F", "GCC_ARM", program)
+
+    config_text = program.files.cmake_config_file.read_text()
+
+    assert "ENABLE_NEWLINES" in config_text
+
+
+def test_config_param_from_target_processed_with_default_name_mangling(program):
+    generate_config("K64F", "GCC_ARM", program)
+
+    config_text = program.files.cmake_config_file.read_text()
+
+    assert "MBED_CONF_TARGET_XIP_ENABLE=0" in config_text
+
+
+def test_macros_from_lib_collected(program):
+    macros = ["NS_USE_EXTERNAL_MBED_TLS"]
+    create_mbed_lib_json(program.mbed_os.root / "connectivity" / "mbed_lib.json", "nanostack", macros=macros)
+
+    generate_config("K64F", "GCC_ARM", program)
+
+    config_text = program.files.cmake_config_file.read_text()
+
+    for macro in macros:
+        assert macro in config_text
+
+
+def test_macros_from_target_collected(program):
+    generate_config("K64F", "GCC_ARM", program)
+
+    config_text = program.files.cmake_config_file.read_text()
+
+    for macro in TARGET_DATA["macros"]:
+        assert macro in config_text
+
+
+def test_target_labels_collected_as_defines(program):
+    generate_config("K64F", "GCC_ARM", program)
+
+    config_text = program.files.cmake_config_file.read_text()
+
+    for label in TARGET_DATA["labels"] + TARGET_DATA["extra_labels"]:
+        assert f"TARGET_{label}" in config_text
+
+    for feature in TARGET_DATA["features"]:
+        assert f"FEATURE_{feature}=1" in config_text
+
+    for device in TARGET_DATA["device_has"]:
+        assert f"DEVICE_{device}=1" in config_text
+
+    for form_factor in TARGET_DATA["supported_form_factors"]:
+        assert f"TARGET_FF_{form_factor}" in config_text
+
+
+def test_overrides_lib_config_param_from_app(matching_target_and_filter, program):
```

Comment:
A few more test ideas:
- Does a lib_config override a target config?
- Does a lib_config override another, previously parsed, lib_config?
- Are separate settings from multiple lib_configs all included?

Reply:
We have an issue where lib configs can override other lib configs, the order of overrides being undefined if this is attempted in multiple libs. The old tools raised an error in that situation. I will add a test to check we raise an error and mark it as xfail until the bug is fixed (I will also raise a defect to track it).

Reply:
Thanks for raising https://github.com/ARMmbed/mbed-tools/issues/155